rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp

```diff
@@ -415,20 +419,105 @@ rmw_ret_t rmw_context_impl_s::shutdown()
 ///=============================================================================
 bool rmw_context_impl_s::is_shutdown() const
 {
-  std::lock_guard<std::mutex> lock(data_->mutex_);
+  std::lock_guard<std::recursive_mutex> lock(data_->mutex_);
   return data_->is_shutdown_;
 }
 
 ///=============================================================================
 bool rmw_context_impl_s::session_is_valid() const
 {
-  std::lock_guard<std::mutex> lock(data_->mutex_);
+  std::lock_guard<std::recursive_mutex> lock(data_->mutex_);
   return z_check(data_->session_);
 }
 
 ///=============================================================================
 std::shared_ptr<rmw_zenoh_cpp::GraphCache> rmw_context_impl_s::graph_cache()
 {
-  std::lock_guard<std::mutex> lock(data_->mutex_);
+  std::lock_guard<std::recursive_mutex> lock(data_->mutex_);
   return data_->graph_cache_;
 }
+
+///=============================================================================
+bool rmw_context_impl_s::create_node_data(
+  const rmw_node_t * const node,
+  const std::string & ns,
+  const std::string & node_name)
+{
+  std::lock_guard<std::recursive_mutex> lock(data_->mutex_);
+  auto node_insertion = data_->nodes_.insert(std::make_pair(node, nullptr));
+  if (!node_insertion.second) {
+    // Node already exists.
+    return false;
+  }
```

Comment:
I might suggest we do this in a different way.  In particular, I think we should delay the insertion of the node_data into the graph until we have done the rest of the construction.  Thus, I would replace this with a check:

```suggestion
  if (data_->nodes_.count(node) > 0) {
    // Node already exists.
    return false;
  }
```

And then down below (where we currently do the `std::make_shared<rmw_zenoh_cpp::NodeData`), we do:

```
data_->nodes_.insert(std::make_pair(node, std::make_shared<NodeData>(std::move(node_id), std::move(entity), std::move(token)));
```

Reply:
Done https://github.com/ros2/rmw_zenoh/pull/259/commits/9d3274d0418d8766ee60112eba702da917176368